### PR TITLE
fix(hubble): retry fetch meta 3 times

### LINKED ISSUE #4535: k8s yaml creation sometimes fails due to hubble 502
Please implement a retry mechanism to avoid crashes in case one request to hubble fails.
image

### PR DESCRIPTION
**Pull request by delgermurun**
Fixes #4535

### CODE REVIEW

File: jina/hubble/hubio.py [hunk 1]
```diff
@@ -588,11 +587,23 @@ def fetch_meta(
...
-            resp.raise_for_status()
+
+        logger = JinaLogger('HubIO.fetch_meta')
+        for i in range(3):
+            try:
+                resp = requests.post(pull_url, json=payload, headers=req_header)
+                if resp.status_code != 200:
+                    if resp.text:
+                        raise Exception(resp.text)
+                    resp.raise_for_status()
```
Comment (numb3r3) [inline jina/hubble/hubio.py:598]: can we wrap the retry as a function, and can reuse this function in some other places.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jina/hubble/hubio.py [hunk 2]
```diff
@@ -588,11 +587,23 @@ def fetch_meta(
...
+                break
+            except Exception as e:
+                logger.debug(
+                    f'Fetching meta info from Jina Hub failed, retry attempt {i+1}/3. '
+                    f'Error: {e!r}'
+                )
+                if i == 2:
```
Comment (JoanFM) [inline jina/hubble/hubio.py:605]: let's make a variable NUM_RETRY

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/hubble/hubio.py (hunk 1 above)
Comment (delgermurun) [reply jina/hubble/hubio.py]: Created `retry` decorator in `helpers`.

The exact failure/retry login in `fetch_meta` doesn't apply to other `HubIO` methods.
So I created a new function inside `fetch_meta`.

File: jina/hubble/hubio.py (hunk 2 above)
Comment (delgermurun) [reply jina/hubble/hubio.py]: Created `retry` decorator in `helpers` with `num_retry` argument.

**Review by JoanFM — DISMISSED** [review]

**Comment by mapleeit** [comment]
@delgermurun Could you rebase again? Tobis said rebase will fix the known black-check in the CI.

**Comment by delgermurun** [comment]
I rebased, but now checks don't start.

**Comment by jacobowitz** [comment]
@delgermurun 
This happens when the devbot adds a commit, closing and reopening the PR mitigates that issue. No idea why this happens in this case.
Tests are running now, but black complains about `helper.py`. I think you need to reformat that file with black

**Comment by delgermurun** [comment]
@mapleeit every checks passed.

@jacobowitz thanks for the help!

**Review by jacobowitz — APPROVED** [review]

**Pull request merged**